File: jira/backend/src/clients/jira/jira-client.ts
import { adf } from "./adf";
import { JIRA_PROJECT_KEY } from "./field-config";

export type JiraOption = { id: string; value?: string; name?: string };
export type JiraIssue = { key: string; fields: Record<string, unknown> };
export type JiraFieldMeta = { allowedValues?: JiraOption[] };
export type JiraTransition = { id: string; name: string; to: { name: string } };
export type JiraWorkType = { id: string; name: string };
export type JiraIssueUpdate = {
	fields?: Record<string, unknown>;
	update?: Record<string, unknown>;
};

export class JiraApiError extends Error {
	readonly status: number;

	constructor(status: number, method: string, path: string, body: string) {
		super(`${method} ${path} -> ${status}: ${body.slice(0, 500)}`);
		this.name = "JiraApiError";
		this.status = status;
	}
}

export interface JiraClient {
	searchIssues(jql: string, fields: readonly string[]): Promise<JiraIssue[]>;
	getIssue(key: string, fields: readonly string[]): Promise<JiraIssue>;
	getEditMeta(key: string): Promise<Record<string, JiraFieldMeta>>;
	updateIssue(key: string, input: JiraIssueUpdate): Promise<void>;
	addComment(key: string, text: string, internal?: boolean): Promise<void>;
	getTransitions(key: string): Promise<JiraTransition[]>;
	transitionIssue(
		key: string,
		transitionId: string,
		fields?: Record<string, unknown>,
	): Promise<void>;
	getWorkTypes(): Promise<JiraWorkType[]>;
}

export type JiraClientConfig = {
	baseUrl: string;
	email: string;
	apiToken: string;
};

const MAX_ATTEMPTS = 6;

export function createRealJiraClient(config: JiraClientConfig): JiraClient {
	const base = config.baseUrl.replace(/\/+$/, "");
	const authorization = `Basic ${btoa(`${config.email}:${config.apiToken}`)}`;

	async function call(
		method: string,
		path: string,
		options: { body?: unknown; params?: Record<string, string> } = {},
	): Promise<unknown> {
		const url = new URL(base + path);
		for (const [k, v] of Object.entries(options.params ?? {})) {
			url.searchParams.set(k, v);
		}
		const init: RequestInit = {
			method,
			headers: {
				Authorization: authorization,
				Accept: "application/json",
				"Content-Type": "application/json",
			},
		};
		if (options.body !== undefined) init.body = JSON.stringify(options.body);

		for (let attempt = 0; attempt < MAX_ATTEMPTS; attempt++) {
			const res = await fetch(url, init);
			if (res.status === 429 || res.status >= 500) {
				const retryAfter =
					Number(res.headers.get("Retry-After")) || 2 ** attempt;
				await Bun.sleep(retryAfter * 1000);
				continue;
			}
			const text = await res.text();
			if (!res.ok) throw new JiraApiError(res.status, method, path, text);
			return text ? JSON.parse(text) : null;
		}
		throw new JiraApiError(503, method, path, "retries exhausted");
	}

	return {
		async searchIssues(jql, fields) {
			const issues: JiraIssue[] = [];
			let nextPageToken: string | undefined;
			do {
				const params: Record<string, string> = {
					jql,
					fields: fields.join(","),
					maxResults: "50",
				};
				if (nextPageToken) params.nextPageToken = nextPageToken;
				const data = (await call("GET", "/rest/api/3/search/jql", {
					params,
				})) as {
					issues?: JiraIssue[];
					nextPageToken?: string;
				};
				issues.push(...(data.issues ?? []));
				nextPageToken = data.nextPageToken;
			} while (nextPageToken);
			return issues;
		},

		async getIssue(key, fields) {
			return (await call(
				"GET",
				`/rest/api/3/issue/${encodeURIComponent(key)}`,
				{
					params: { fields: fields.join(",") },
				},
			)) as JiraIssue;
		},

		async getEditMeta(key) {
			const data = (await call(
				"GET",
				`/rest/api/3/issue/${encodeURIComponent(key)}/editmeta`,
			)) as {
				fields: Record<string, JiraFieldMeta>;
			};
			return data.fields;
		},

		async updateIssue(key, input) {
			await call("PUT", `/rest/api/3/issue/${encodeURIComponent(key)}`, {
				body: input,
			});
		},

		async addComment(key, text, internal = false) {
			const body: Record<string, unknown> = { body: adf(text) };
			if (internal) {
				// Jira Service Management reads this property to keep the comment agent-only
				body.properties = [
					{ key: "sd.public.comment", value: { internal: true } },
				];
			}
			await call(
				"POST",
				`/rest/api/3/issue/${encodeURIComponent(key)}/comment`,
				{ body },
			);
		},

		async getTransitions(key) {
			const data = (await call(
				"GET",
				`/rest/api/3/issue/${encodeURIComponent(key)}/transitions`,
			)) as { transitions: JiraTransition[] };
			return data.transitions;
		},

		async transitionIssue(key, transitionId, fields) {
			const body: Record<string, unknown> = {
				transition: { id: transitionId },
			};
			if (fields) body.fields = fields;
			await call(
				"POST",
				`/rest/api/3/issue/${encodeURIComponent(key)}/transitions`,
				{ body },
			);
		},

		async getWorkTypes() {
			const data = (await call(
				"GET",
				`/rest/api/3/issue/createmeta/${JIRA_PROJECT_KEY}/issuetypes`,
			)) as { issueTypes?: JiraWorkType[]; values?: JiraWorkType[] };
			return data.issueTypes ?? data.values ?? [];
		},
	};
}

File: backend/jira_scripts/jira_client.py
"""Thin Jira Cloud REST client plus helpers shared by upload.py and loop.py."""

import os
import re
import time

import requests
from dotenv import load_dotenv

import backend.jira_scripts.config as config

load_dotenv()


class Jira:
    def __init__(self):
        self.base = os.environ["JIRA_BASE_URL"].rstrip("/")
        self.s = requests.Session()
        self.s.auth = (os.environ["JIRA_EMAIL"], os.environ["JIRA_API_TOKEN"])
        self.s.headers.update({"Accept": "application/json", "Content-Type": "application/json"})

    def call(self, method, path, **kw):
        """Request with retry on rate limits (429) and server errors."""
        for attempt in range(6):
            r = self.s.request(method, self.base + path, timeout=30, **kw)
            if r.status_code == 429 or r.status_code >= 500:
                wait = int(r.headers.get("Retry-After", 2 ** attempt))
                print(f"  ...Jira said {r.status_code}, retrying in {wait}s")
                time.sleep(wait)
                continue
            if not r.ok:
                raise JiraError(r.status_code, r.text, method, path)
            return r.json() if r.text else None
        raise JiraError(r.status_code, r.text, method, path)

    # ---------- metadata ----------

    def work_types(self):
        data = self.call("GET", f"/rest/api/3/issue/createmeta/{config.PROJECT_KEY}/issuetypes")
        return data.get("issueTypes") or data.get("values") or []

    def create_fields(self, issuetype_id):
        """Fields (with allowedValues) available when creating this work type."""
        data = self.call(
            "GET",
            f"/rest/api/3/issue/createmeta/{config.PROJECT_KEY}/issuetypes/{issuetype_id}",
            params={"maxResults": 200},
        )
        items = data.get("fields") or data.get("values") or []
        return {f["fieldId"]: f for f in items}

    def resolutions(self):
        return self.call("GET", "/rest/api/3/resolution")

    # ---------- issues ----------

    def search(self, jql, fields):
        token = None
        while True:
            params = {"jql": jql, "fields": ",".join(fields), "maxResults": 50}
            if token:
                params["nextPageToken"] = token
            data = self.call("GET", "/rest/api/3/search/jql", params=params)
            yield from data.get("issues", [])
            token = data.get("nextPageToken")
            if not token:
                break

    def create_issue(self, fields):
        return self.call("POST", "/rest/api/3/issue", json={"fields": fields})

    def update_issue(self, key, fields=None, update=None):
        body = {}
        if fields:
            body["fields"] = fields
        if update:
            body["update"] = update
        return self.call("PUT", f"/rest/api/3/issue/{key}", json=body)

    def add_comment(self, key, text, internal=False):
        body = {"body": adf(text)}
        if internal:
            # Jira Service Management reads this property to keep the comment agent-only
            body["properties"] = [{"key": "sd.public.comment", "value": {"internal": True}}]
        return self.call("POST", f"/rest/api/3/issue/{key}/comment", json=body)

    def transitions(self, key):
        return self.call("GET", f"/rest/api/3/issue/{key}/transitions")["transitions"]

    def transition(self, key, transition_id, fields=None):
        body = {"transition": {"id": transition_id}}
        if fields:
            body["fields"] = fields
        return self.call("POST", f"/rest/api/3/issue/{key}/transitions", json=body)


class JiraError(Exception):
    def __init__(self, status, text, method, path):
        super().__init__(f"{method} {path} -> {status}: {text[:500]}")
        self.status = status
        self.text = text


# ---------- Atlassian Document Format ----------

def adf(text):
    """Plain text -> ADF. Blank lines separate paragraphs."""
    paras = [p for p in re.split(r"\n\s*\n", text or "") if p.strip()] or [" "]
    content = []
    for p in paras:
        lines = p.split("\n")
        inline = []
        for i, line in enumerate(lines):
            if i:
                inline.append({"type": "hardBreak"})
            if line:
                inline.append({"type": "text", "text": line})
        content.append({"type": "paragraph", "content": inline})
    return {"type": "doc", "version": 1, "content": content}


def adf_to_text(node):
    """ADF -> plain text (good enough for feeding a model)."""
    if not node:
        return ""
    if isinstance(node, str):
        return node
    t = node.get("type")
    if t == "text":
        return node.get("text", "")
    if t == "hardBreak":
        return "\n"
    inner = "".join(adf_to_text(c) for c in node.get("content", []))
    return inner + ("\n\n" if t == "paragraph" else "")


# ---------- matching our values to Jira's options ----------

# Order matters: "highest"/"lowest" must be checked before "high"/"low".
_RANK_WORDS = [
    (0, ["highest", "critical", "extensive", "widespread", "major"]),
    (4, ["lowest", "no direct", "none", "information", "informational"]),
    (1, ["high", "significant", "large"]),
    (2, ["medium", "moderate", "limited"]),
    (3, ["low", "minor", "localized", "localised"]),
]


def rank(label):
    """0 = most severe ... 4 = least severe. None if unrecognised."""
    s = (label or "").lower()
    for r, words in _RANK_WORDS:
        for w in words:
            if re.search(r"\b" + re.escape(w) + r"\b", s):
                return r
    return None


def option_label(opt):
    return opt.get("value") or opt.get("name") or ""


def match_option(value, allowed):
    """Find the Jira option for `value`: exact name first, then same severity rank,
    then the nearest rank. Returns the option dict or None."""
    if not value or not allowed:
        return None
    v = value.strip().lower()
    for o in allowed:
        if option_label(o).strip().lower() == v:
            return o
    r = rank(value)
    if r is None:
        return None
    ranked = [(rank(option_label(o)), o) for o in allowed]
    ranked = [(x, o) for x, o in ranked if x is not None]
    if not ranked:
        return None
    return min(ranked, key=lambda p: abs(p[0] - r))[1]


def find_work_type(types, name):
    """'Incident' -> the Jira work type whose name contains 'incident'
    (and not 'approvals', to skip 'Service request with approvals')."""
    needle = config.WORK_TYPES.get(name, name).lower()
    for t in types:
        n = t["name"].lower()
        if needle in n and "approval" not in n:
            return t
    return None
